DeepSpeech.py

```diff
@@ -694,594 +651,999 @@ def stopwatch(start_duration=0):
+                if index >= 0:
+                    self._send_answer(str(index))
+                    return
+            elif self.path.startswith(PREFIX_GET_JOB):
+                job = COORD.get_job(worker=int(self.path[len(PREFIX_GET_JOB):]))
+                if job:
+                    self._send_answer(pickle.dumps(job))
+                    return
+            self.send_response(404)
+            self.end_headers()
+
+        def do_POST(self):
+            src = self.rfile.read(int(self.headers['content-length']))
+            job = COORD.next_job(pickle.loads(src))
+            self._send_answer(pickle.dumps(job))
+
+        def log_message(self, format, *args):
+            '''Overriding base method to suppress web handler messages on stdout.
+            '''
+            return
+
+
+    def __init__(self):
+        ''' Central training coordination class.
+        Used for distributing jobs among workers of a cluster.
+        Instantiated on all workers, calls of non-chief workers will transparently
+        HTTP-forwarded to the chief worker instance.
+        '''
+        self._init()
+        self._lock = Lock()
+        if is_chief:
+            self._httpd = BaseHTTPServer.HTTPServer((FLAGS.coord_host, FLAGS.coord_port), TrainingCoordinator.TrainingCoordinationHandler)
+
+    def _reset_counters(self):
+        self._index_train = 0
+        self._index_dev = 0
+        self._index_test = 0
+
+    def _init(self):
+        self._epochs_running = []
+        self._epochs_done = []
+        self._reset_counters()
+
+    def _log_all_jobs(self):
+        '''Use this to debug-print epoch state'''
+        log_debug('Epochs - running: %d, done: %d' % (len(self._epochs_done), len(self._epochs_running)))
+        for epoch in self._epochs_running:
+            log_debug('       - running: ' + epoch.job_status())
+
+    def start_coordination(self, data_sets, step=0):
+        '''Starts to coordinate epochs and jobs among workers on base of
+        data-set sizes, the (global) step and FLAGS parameters.
+
+        Args:
+            data_sets (DataSets): data-sets to be used for coordinated training
+
+        Kwargs:
+            step (int): global step of a loaded model to determine starting point
+        '''
+        with self._lock:
+            self._init()
+
+            # Number of GPUs per worker - fixed for now by local reality or cluster setup
+            gpus_per_worker = len(available_devices)
+
+            # Number of batches processed per job per worker
+            batches_per_job  = gpus_per_worker * max(1, FLAGS.iters_per_worker)
+
+            # Number of batches per global step
+            batches_per_step = gpus_per_worker * max(1, FLAGS.replicas_to_agg)
+
+            # Number of global steps per epoch - to be at least 1
+            steps_per_epoch = max(1, data_sets.train.total_batches / batches_per_step)
+
+            # The start epoch of our training
+            self._epoch = step / steps_per_epoch
```

Comment:
This needs to be `step // steps_per_epoch` (integer division) for Python 2/3 compat. In Python 3, this will evaluate to a float value if `steps_per_epoch` doesn't divide `step`.